backend/push.js

```diff
@@ -0,0 +1,19 @@
+const pushInstance = require('web-push')
+const pushSubscriptions: Map<string, any> = new Map()
```

Comment:
Hi @snowteamer, 
Can I borrow your opinion for what Greg suggested in the above comment please?
(It's my first time doing a task that involves changes in the back-end area and need some help)

`pushSubscriptions` map here stores push-subscription data that users send  as `{ subscriptionId` : `subscription-data }` format and the `pushInstance` above it uses these data to send push notification.

Currently all push notification related logics on the server side are implemented as corresponding REST API (`/push/*` routes in `routes.js`) but what @taoeffect suggests is to integrate these into the existing pubsub system.

I will have to take some more time to study/figure out how the pubsub works, but would you have some advice on how I can approach this task?

Thanks,

Reply:
The current pubsub system is somewhat contract-specific (I could work on that if requested), so there would be a few caveats

I will try to summarize how to extend the default protocol. Both the client and server can receive and send JSON messages.

To add custom WebSocket message to the pubsub protocol, two steps are needed:

### Server-side
Custom message handlers can be registered by adding them as entries of the `messageHandlers` option in the `createServer` call in `server.js`.

For example, a `push-sub` message handler will be called when the server receives a `push-sub` message.
To implement it, you would have to update the following code in `server.js`:
https://github.com/okTurtles/group-income/blob/3200d2534e192a0c29b1e8901d1ba740b259dfa2/backend/server.js#L88-L95

Then it would become something like:
```js
sbp('okTurtles.data/set', PUBSUB_INSTANCE, createServer(hapi.listener, {
  messageHandlers: {
    [ PUSH_SUB ] (message: Message) {
      // Your handler code here.
    }
  },
  serverHandlers: {
    connection (socket: Object, request: Object) {
      const versionInfo = { GI_VERSION, CONTRACTS_VERSION }
      socket.send(createNotification(NOTIFICATION_TYPE.VERSION_INFO, versionInfo))
    }
  }
}))
```

A few notes about message handlers:
-  Their `this` binding refers to the connected `ws` socket object. See https://github.com/websockets/ws.
    That socket can be used e.g. to send messages back to the connected client. A few helpers like `createMessage`, `createNotification` and `createErrorResponse` are available to create such messages. However note that in that case, the client must have matching handlers. For example, if the server sends a `foo` message back to the client, the client must have a `foo` message handler.
    
### Messages
- Messages are ordinary JSON objects, albeit with a required read-only `type: string` field, but no `id` field like in some other implementations. I would add in a more generic system, but for un/subscribing to contracts a `contractID` field was enough.

### Client-side: sending
In the browser, the pubsub client is created in chelonia/connect':
https://github.com/okTurtles/group-income/blob/3200d2534e192a0c29b1e8901d1ba740b259dfa2/shared/domains/chelonia/chelonia.js#L141
You can see it has a `messageHandlers` option like its server-side counterpart.

The client wraps a `socket: WebSocket` property, which can be used to send data.

Besides, the client API is quite minimalist and has not much beyond `pub`, `sub` and `unsub`.
- Currently, the client `pub` method is not implemented, as it was not required to publish contract events a.k.a. GI messages to the server (that was already done via POST requests).
- If the client has to send custom messages, this should be possible using `client.socket.send()`, or `this.socket.send()` from within message handlers. Helpers `createMessage` and `createRequest` are available.
- The default `sub` client method likely won't work to subscribe to anything else than GI contracts, since if you call `sub(subscriptionId)` then the current code will try to sync a non-existent `subscriptionId` contract upon reconnection, because of this code in the client socket `open` event handler:
    https://github.com/okTurtles/group-income/blob/3200d2534e192a0c29b1e8901d1ba740b259dfa2/shared/pubsub.js#L330-L334

Reply:
@taoeffect Regarding your suggestion above about moving the push subscription map into the pubsub system, Did you have a specific place in your mind when you were mentioning that?

I was thinking that 
[here](https://github.com/okTurtles/group-income/blob/3200d2534e192a0c29b1e8901d1ba740b259dfa2/backend/pubsub.js#L94C1-L94C1) in `createServer()` of `backend/pubsub.js` (adding `server.pushSubscriptions = new Map()` there) 
would be ideal, because that way this map can be referenced easily in any server-side message handler via something like `this.server.pushSubscriptions`. Does it sound good to you too?

Reply:
@SebinSong I'm not sure. When I've used pubsub systems before, they've worked really well with the concept of "rooms" - or in other words, identifiers/channels that you subscribe to, that are identified by arbitrary strings of the developer's choosing. Maybe we can use this opportunity to make the existing pubsub system we have more generic? Then all of the subscriptions can be done on the same map potentially.